From bd221d7f2c65cc780d7b636543c950c63a168230 Mon Sep 17 00:00:00 2001
From: Romain Rigaux <romain@cloudera.com>
Date: Mon, 13 Oct 2014 17:47:08 -0700
Subject: [PATCH 0039/1173] HUE-2348 [search] Highlight terms inplace in the
 field

Return the full fragment when matching
Fix missing matches when ids are not string
Escape highlighted HTML
---
 apps/search/src/search/models.py        |   12 ++++++++++--
 desktop/libs/libsolr/src/libsolr/api.py |    3 ++-
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/apps/search/src/search/models.py b/apps/search/src/search/models.py
index 35eba7f..bf00eac 100644
--- a/apps/search/src/search/models.py
+++ b/apps/search/src/search/models.py
@@ -540,8 +540,16 @@ def augment_solr_response(response, collection, query):
     id_field = collection.get('idField')
     if id_field:
       for doc in response['response']['docs']:
-        if id_field in doc and doc[id_field] in highlighted_fields:
-          doc.update(response['highlighting'][doc[id_field]])
+        if id_field in doc and str(doc[id_field]) in highlighted_fields:
+          highlighting = response['highlighting'][str(doc[id_field])]
+
+          if highlighting:
+            escaped_highlighting = {}
+            for field, hls in highlighting.iteritems():
+              _hls = [escape(smart_unicode(hl, errors='replace')).replace('&lt;em&gt;', '<em>').replace('&lt;/em&gt;', '</em>') for hl in hls]
+              escaped_highlighting[field] = _hls
+  
+            doc.update(escaped_highlighting)
     else:
       response['warning'] = _("The Solr schema requires an id field for performing the result highlighting")
 
diff --git a/desktop/libs/libsolr/src/libsolr/api.py b/desktop/libs/libsolr/src/libsolr/api.py
index 675b830..00c1174 100644
--- a/desktop/libs/libsolr/src/libsolr/api.py
+++ b/desktop/libs/libsolr/src/libsolr/api.py
@@ -165,7 +165,8 @@ class SolrApi(object):
     params += (
       ('hl', 'true'),
       ('hl.fl', '*'),
-      ('hl.snippets', 3)
+      ('hl.snippets', 3),
+      ('hl.fragsize', 0),
     )
 
     if collection['template']['fieldsSelected']:
-- 
1.7.9.5

